core/trino-main/src/main/java/io/trino/sql/planner/LocalDynamicFilterConsumer.java

```diff
@@ -49,39 +52,58 @@
 
     private final SettableFuture<TupleDomain<DynamicFilterId>> resultFuture;
 
-    // Number of build-side partitions to be collected.
-    private final int partitionCount;
+    // Number of build-side partitions to be collected, must be provided by setPartitionCount
+    @GuardedBy("this")
+    private int expectedPartitionCount = PARTITION_COUNT_INITIAL_VALUE;
 
     // The resulting predicates from each build-side partition.
+    @GuardedBy("this")
     private final List<TupleDomain<DynamicFilterId>> partitions;
 
-    public LocalDynamicFilterConsumer(Map<DynamicFilterId, Integer> buildChannels, Map<DynamicFilterId, Type> filterBuildTypes, int partitionCount)
+    public LocalDynamicFilterConsumer(Map<DynamicFilterId, Integer> buildChannels, Map<DynamicFilterId, Type> filterBuildTypes)
     {
         this.buildChannels = requireNonNull(buildChannels, "buildChannels is null");
         this.filterBuildTypes = requireNonNull(filterBuildTypes, "filterBuildTypes is null");
         verify(buildChannels.keySet().equals(filterBuildTypes.keySet()), "filterBuildTypes and buildChannels must have same keys");
 
         this.resultFuture = SettableFuture.create();
-
-        this.partitionCount = partitionCount;
-        this.partitions = new ArrayList<>(partitionCount);
+        this.partitions = new ArrayList<>();
     }
 
     public ListenableFuture<Map<DynamicFilterId, Domain>> getDynamicFilterDomains()
     {
         return Futures.transform(resultFuture, this::convertTupleDomain, directExecutor());
     }
 
-    private void addPartition(TupleDomain<DynamicFilterId> tupleDomain)
+    @Override
+    public void addPartition(TupleDomain<DynamicFilterId> tupleDomain)
     {
         TupleDomain<DynamicFilterId> result = null;
         synchronized (this) {
             // Called concurrently by each DynamicFilterSourceOperator instance (when collection is over).
-            verify(partitions.size() < partitionCount);
+            verify(expectedPartitionCount == PARTITION_COUNT_INITIAL_VALUE || partitions.size() < expectedPartitionCount);
             // NOTE: may result in a bit more relaxed constraint if there are multiple columns and multiple rows.
             // See the comment at TupleDomain::columnWiseUnion() for more details.
             partitions.add(tupleDomain);
-            if (partitions.size() == partitionCount || tupleDomain.isAll()) {
+            if (partitions.size() == expectedPartitionCount || tupleDomain.isAll()) {
+                // No more partitions are left to be processed.
+                result = TupleDomain.columnWiseUnion(partitions);
+            }
+        }
+
+        if (result != null) {
+            resultFuture.set(result);
+        }
+    }
+
+    @Override
+    public void setPartitionCount(int partitionCount)
+    {
+        TupleDomain<DynamicFilterId> result = null;
+        synchronized (this) {
+            verify(expectedPartitionCount == PARTITION_COUNT_INITIAL_VALUE);
```

Comment:
replace with checkState (as precondition can be violated by the caller) and provide exception message.